pkg/csi/controller_server.go

```diff
@@ -457,31 +663,85 @@ func getVolumeCapabilityAccessModes(vc []csi.VolumeCapability_AccessMode_Mode) [
 	return vca
 }
 
-func (cs *ControllerServer) waitForPVCState(name string, stateDescription string,
-	predicate func(pvc *corev1.PersistentVolumeClaim) bool) bool {
-	timer := time.NewTimer(timeoutAttachDetach)
-	defer timer.Stop()
-	timeout := timer.C
+func waitForVolSettled(lhClient *lhclientset.Clientset, lhVolName, nodeID string) (bool, error) {
+	volume, err := lhClient.LonghornV1beta2().Volumes(LonghornNS).Get(context.TODO(), lhVolName, metav1.GetOptions{})
+	if err != nil {
+		logrus.Warnf("waitForVolumeSettled: error while waiting for volume %s to be settled. Err: %v", lhVolName, err)
+		return false, err
+	}
+	// check attached correctly or detached
+	if volAttachedCorrectly(volume, nodeID) {
+		return true, nil
+	}
+	if volume.Status.State == lhv1beta2.VolumeStateDetached {
+		return true, nil
+	}
```

Comment:
The comment is a little confusing, IIUC, `volAttachedCorrectly()` is only for checking volume attachment.

Reply:
updated, thanks!